Output partial results

benv666 opened the pull request:
This PR keeps track of the longest matching prefix found so far and adds some output to make it useful.
The purpose of this is so I can set a way too long prefix and have my machine crunch away at it until I actually need it and have the most beautiful prefix without worrying if it'll ever find it in time.
E.g.
```
./mine 0xa77E57EDF00d1234
Searching for 0xa77E57EDF00d1234 using 8 threads
Partial prefix match of len 1: Address: 0xad429891ff2121d47a3dfb3fec54efedf412a754 - salt: 0x04
Partial prefix match of len 2: Address: 0xa742c9391d131096e984d26620b4d8eea3dd37a9 - salt: 0x0307
Partial prefix match of len 3: Address: 0xa772f85bd0399b0d8627494c8a68a91c62d32d4d - salt: 0x2ff3
Partial prefix match of len 4: Address: 0xa77e39c1ebeac0fbd14acbf260f10510a3bab466 - salt: 0xe2ff
Partial prefix match of len 5: Address: 0xa77e54711ef71718cb73e3013c0520be4cb90362 - salt: 0x08401b
Partial prefix match of len 6: Address: 0xa77e577f7a0f48e1058a629cdde60f077b968139 - salt: 0xaf11ec
At salt 0x03d4ccc0... 5.00s (12.86M salts/sec), longest prefix found matches 6 chars using salt 0xaf11ec giving address: 0xa77e577f7a0f48e1058a629cdde60f077b968139
Partial prefix match of len 7: Address: 0xa77e57ef6eaa03518ee1ea68ec887e3f8d25e441 - salt: 0x04ecdd65
At salt 0x07ade080... 10.00s (12.91M salts/sec), longest prefix found matches 7 chars using salt 0x04ecdd65 giving address: 0xa77e57ef6eaa03518ee1ea68ec887e3f8d25e441
At salt 0x0bf1fc80... 15.00s (14.31M salts/sec), longest prefix found matches 7 chars using salt 0x04ecdd65 giving address: 0xa77e57ef6eaa03518ee1ea68ec887e3f8d25e441
At salt 0x0fd94740... 20.00s (13.10M salts/sec), longest prefix found matches 7 chars using salt 0x04ecdd65 giving address: 0xa77e57ef6eaa03518ee1ea68ec887e3f8d25e441
At salt 0x13f39980... 25.00s (13.77M salts/sec), longest prefix found matches 7 chars using salt 0x04ecdd65 giving address: 0xa77e57ef6eaa03518ee1ea68ec887e3f8d25e441
```
Now I can abort it and use the last found prefix instead of having to restart/continue and hoping it'll win the lottery in time.

NOTE: Performance wise it didn't seem to impact the ~13M-ish salts/sec on my machine, but perhaps this can be tuned :)

jshufro commented:
Can we put this behavior behind a flag? by the way you can pass a starting salt to 'resume'

benv666 commented:
Flag is certainly possible, I'll see if I can add that later today. (assuming you mean a runtime flag)

Also the resume option is known, my point with restarting is that if you now find out your salt takes too long to compute, you'll need to restart with a smaller salt without my patch. With my patch you can just pick the longest found prefix without having to restart.

benv666 commented:
@jshufro Thoughts? :⁠-⁠)

jshufro commented on main.c:
```diff
@@ -164,6 +168,39 @@ print_salt(uint64_t salt)
+
+	// optimization: check if the longest + 1 nibble matches, if not we can immediately stop
+	unsigned char cmp;
+	if (longest_prefix_len & 1) { // uneven -> next one should be in last nibble
```
thread is accessing a global here- not safe

jshufro commented on main.c:
```diff
@@ -205,6 +246,19 @@ iteration(unsigned char *state, const unsigned char *prefix, unsigned char *phas
 			return true;
 		}
+		if (intermediate_output && (my_prefix_len > longest_prefix_len)) {
+			longest_prefix_salt = salt + j;
```
cross-thread write contention here

jshufro commented on main.c:
```diff
@@ -205,6 +246,19 @@ iteration(unsigned char *state, const unsigned char *prefix, unsigned char *phas
 		}
+		if (intermediate_output && (my_prefix_len > longest_prefix_len)) {
+			longest_prefix_salt = salt + j;
+			memcpy(longest_prefix_addr, addr, ADDRESS_BYTES);
```
extremely unsafe memcpy

jshufro reviewed (changes requested):
Alright, on deeper review, this is intolerably thread-unsafe. I can't merge it as-is, however, I like the idea.

That said, I don't like how much complexity it adds to the already-spaghetti-ish code that I slapped out in a weekend.

Stand by while I implement "plugin" support... thread safety is something I can't bake into it, though, since plugins will want to observe every mined address, and they exist in an arena ephemerally during the thread's main loop... so be prepared to learn the agony of systems engineering if you want to follow up on this

jshufro commented:
Plugin support was added https://github.com/jshufro/rocketpool-vfvm/commit/7559a02904024bb7e2f83f35ba7dba50dfa17099

benv666 commented:
I'm giving it a shot rewriting it with your new plugin system while trying to keep thread-safety in mind, but while tinkering I noticed the new `on_iteration` callback does not provide the plugin with the corresponding salt.

The intermediate output does need it, since that's the entire point of this plugin.
Can we add salt as third component to the `on_iteration` callback?

jshufro commented:
yeah feel free to add whatever fields you need

benv666 commented:
I made a stab, pulled it through `-fsanitize=thread` and `-fsanitize=leak` and that looked reasonable to me, but constructive criticism is very welcome :)

Another thing that crossed my mind would be to install an interrupt handler and print the partial results on CTRL-C, any thoughts on that?

benv666 commented:
@jshufro Once again I ask for your input :⁠-⁠)

jshufro commented on plugins/partial-output/partial-output.h:
Remove this file

benv666 replied:
done

jshufro commented on plugins/partial-output/partial-output.c:
```diff
@@ -0,0 +1,187 @@
+	memset(target_prefix_addr, 0, ADDRESS_BYTES);
+	memcpy(target_prefix_addr,prefix,prefix_len);
+
+	// Set up signal handler for SIGINT (ctrl-c)
+	struct sigaction sa;
+	sa.sa_handler = sigint_handler;
+	sigemptyset(&sa.sa_mask);
+	sa.sa_flags = 0;
+	sigaction(SIGINT, &sa, NULL);
```
this needs to be moved to main.c and call a new plugin->on_interrupt

jshufro commented on plugins/partial-output/partial-output.c:
```diff
@@ -0,0 +1,187 @@
+	const size_t my_prefix_len = get_prefix_len(addr);
+
+	if (my_prefix_len > longest_prefix_len) {
+		pthread_mutex_lock(&match_mutex);
```
what's the performance like with these locks? there's probably a lock-free solution

benv666 replied:
Since `my_prefix_len` only increases when a new longest match is found, this is imho not an issue. Only the first few matches will happen within seconds of starting the program, after that it'll be minutes/hours/days between this lock is used.

TDLR: negligible impact.

jshufro commented on plugins/partial-output/partial-output.c:
```diff
@@ -0,0 +1,187 @@
+static void
+on_progress(void)
+{
+	if (longest_prefix_len) {
+		pthread_mutex_lock(&match_mutex);
+		PRINTF("Longest prefix found matches %ld chars using salt ", longest_prefix_len);
+		print_salt(longest_prefix_salt);
+		printf(" giving address: 0x");
+		for (size_t i = 0; i < ADDRESS_BYTES; i++)
+			printf("%02x", longest_prefix_addr[i]);
+		printf("\n");
+		pthread_mutex_unlock(&match_mutex);
+	}
```
```suggestion
	if (longest_prefix_len == 0) {
		return;
	}

	pthread_mutex_lock(&match_mutex);
	PRINTF("Longest prefix found matches %ld chars using salt ", longest_prefix_len);
	print_salt(longest_prefix_salt);
	printf(" giving address: 0x");
	for (size_t i = 0; i < ADDRESS_BYTES; i++)
		printf("%02x", longest_prefix_addr[i]);
	printf("\n");
	pthread_mutex_unlock(&match_mutex);
```

benv666 commented:
@jshufro Thanks again for the constructive criticism.

I've (hopefully) addressed your concerns, and possible introduced some new ones with the new `on_interrupt` construct I've created.

Note that I've opted to raise() the old signal in the signal handler, after tinkering a bit with setting the `done` bool to true in the sigint handler and testing with locks and their performance (including a construct which raised a `interrupted` sig_atomic_t causing the done to be toggled after a progress report, causing a 5-10 sec delay before quit). I feel like this has the best user experience, but feedback is again welcome.

jshufro commented on plugins/partial-output/partial-output.c:
```diff
@@ -0,0 +1,182 @@
+{
+	const size_t my_prefix_len = get_prefix_len(addr);
+
+	if (my_prefix_len > longest_prefix_len) {
+		pthread_mutex_lock(&match_mutex);
+
+		longest_prefix_salt = salt;
+		memcpy(longest_prefix_addr, addr, ADDRESS_BYTES);
+		atomic_store(&longest_prefix_len, my_prefix_len);
+
+		pthread_mutex_unlock(&match_mutex);
+	}
```
```suggestion
	if (my_prefix_len <= longest_prefix_len) {
		return;
	}
	pthread_mutex_lock(&match_mutex);

	longest_prefix_salt = salt;
	memcpy(longest_prefix_addr, addr, ADDRESS_BYTES);
	atomic_store(&longest_prefix_len, my_prefix_len);

	pthread_mutex_unlock(&match_mutex);
```

not trying to be difficult, but early returns make for nicer c

benv666 replied:
Agreed, fixed

jshufro reviewed (changes requested):
I'm impressed you stuck with it all the way to this point. It's in really good shape now, just one nitpick.

Also, if you don't mind, before we merge, can you rewrite history into 2 commits?

1 which updates main.c, plugin.h and plugins/examples, and a second commit which adds partial-output?

benv666 committed:
Add on_interrupt feature, add salt to on_iteration

benv666 committed:
Add partial output plugin

benv666 commented:
@jshufro Other than your requests no new changes, as far as I'm concerned this can be merged.
Thanks for the feedback and suggestions!
